src/iterators/inverted_index_iterator.c

```diff
@@ -60,6 +61,118 @@ size_t InvIndIterator_NumEstimated(QueryIterator *base) {
+    // The numeric tree was either completely deleted or a node was split or removed.
+    // The cursor is invalidated.
+    base->isAborted = true;
+    return VALIDATE_ABORTED;
+  }
+
+  return VALIDATE_OK;
+}
+
+static ValidateStatus TermCheckAbort(QueryIterator *base) {
+  InvIndIterator *it = (InvIndIterator *)base;
+  if (!it->sctx) {
+    return VALIDATE_OK;
+  }
+  InvertedIndex *idx = Redis_OpenInvertedIndex(it->sctx, base->current->data.term.term->str,
+      base->current->data.term.term->len, false, NULL);
+  if (!idx || idx == TRIEMAP_NOTFOUND || it->idx != idx) {
+    // The inverted index was collected entirely by GC.
+    // All the documents that were inside were deleted and new ones were added.
+    // We will not continue reading those new results and instead abort reading
+    // for this specific inverted index.
+    base->isAborted = true;
+    return VALIDATE_ABORTED;
+  }
+  return VALIDATE_OK;
+}
+
+static ValidateStatus TagCheckAbort(QueryIterator *base) {
+  TagInvIndIterator *it = (TagInvIndIterator *)base;
+  if (!it->base.sctx) {
+    return VALIDATE_OK;
+  }
+  size_t sz;
+  InvertedIndex *idx = TagIndex_OpenIndex(it->tagIdx, base->current->data.term.term->str,
+      base->current->data.term.term->len, false, &sz);
+  if (!idx || idx == TRIEMAP_NOTFOUND || it->base.idx != idx) {
+    // The inverted index was collected entirely by GC.
+    // All the documents that were inside were deleted and new ones were added.
+    // We will not continue reading those new results and instead abort reading
+    // for this specific inverted index.
+
+    base->isAborted = true;
+    return VALIDATE_ABORTED;
+  }
+  return VALIDATE_OK;
+}
+
+static ValidateStatus InvIndIterator_Revalidate(QueryIterator *base) {
+  // Here we should apply the specifics of Term, Tag and Numeric
+
+  InvIndIterator *it = (InvIndIterator *)base;
+  if (base->atEOF) {
+    // Save time and state if we are already at the end
+    return VALIDATE_OK;
+  }
+
+  ValidateStatus ret = it->CheckAbort(it);
+
+  if(ret != VALIDATE_OK) {
+    return ret;
+  }
+
+  // the gc marker tells us if there is a chance the keys has undergone GC while we were asleep
+  if (it->gcMarker == it->idx->gcMarker) {
+    // no GC - we just go to the same offset we were at
+    size_t offset = it->blockReader.buffReader.pos;
+    SetCurrentBlockReader(it);
+    it->blockReader.buffReader.pos = offset;
+    ret = VALIDATE_OK;
+  } else {
+    // if there has been a GC cycle on this key while we were asleep, the offset might not be valid
+    // anymore. This means that we need to seek to last docId we were at
+
+    // keep the last docId we were at
+    t_docId lastDocId = base->lastDocId;
+    // reset the state of the reader
+    base->Rewind(base);
+    IteratorStatus rc = base->SkipTo(base, lastDocId);
+    if (rc == ITERATOR_NOTFOUND) {
+      ret = VALIDATE_MOVED;
+    } else if (rc == ITERATOR_EOF) {
+      ret = VALIDATE_OK;
+    }
+  }
+
+  return ret;
```

Comment:
A bit hard to follow. Consider either assuming `ret == VALIDATE_OK` and only change it if we need to (the `rc == ITERATOR_NOTFOUND` case), or have a clearer default
```suggestion
  if (ret != VALIDATE_OK) {
    return ret;
  }

  // the gc marker tells us if there is a chance the keys have undergone GC while we were asleep
  if (it->gcMarker == it->idx->gcMarker) {
    // no GC - we just go to the same offset we were at
    size_t offset = it->blockReader.buffReader.pos;
    SetCurrentBlockReader(it);
    it->blockReader.buffReader.pos = offset;
  } else {
    // if there has been a GC cycle on this key while we were asleep, the offset might not be valid
    // anymore. This means that we need to seek the last docId we were at

    // keep the last docId we were at
    t_docId lastDocId = base->lastDocId;
    // reset the state of the reader
    base->Rewind(base);
    IteratorStatus rc = base->SkipTo(base, lastDocId);
    if (rc == ITERATOR_NOTFOUND) {
      ret = VALIDATE_MOVED;
    }
  }

  return ret;
```